*WIP* Initial Implementation of kpod launch

mheon opened the pull request:
This is a mostly functional implementation of `kpod launch.` This is mostly an analogue for `kubectl run` which can be run without a running cri-o daemon, with a bit more flexibility. Eventually it will allow you to insert a container into an existing pod for debugging/inspection, but for now it creates new pods and launches containers into them.

The code is working and should be ready to test (with the exception of a number of CLI options that aren't implemented yet). It successfully creates pods and containers, which can be seen and interacted with by the normal CRI-O daemon (it should work without the daemon running as well, but because we don't have a `kpod remove` yet you can't remove containers without CRI-O).

Ongoing work is documented in the TODO list at the top of launch.go.

mheon commented:
CC @mrunalp @rhatdan

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.BoolFlag{
+			Name:  "attach",
+			Usage: "attach to the container once it is created",
```
"attach to the pod once it is created",

mheon replied:
I don't know if this one makes sense - it implies we're attaching to every container in a pod, not just the one created by `kpod launch`. Support for inserting a container in existing pods isn't too far from done, so it seems a little confusing

rhatdan replied:
So this should be 
attach terminal to initial/Primary container once it is created?
Once we have multiple, we will need 
kpod attach/exec and specify the container id?

mheon replied:
"Primary container" sounds like a good way of phrasing it, I think. We can specify that all operations that target a single container will affect only the primary.

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.StringSliceFlag{
+			Name:  "env",
+			Usage: "specify environment variables to be set inside launched container, specified as KEY=VALUE",
```
We want kpod launch --help to use better descriptions then value.
```
			Usage: "specify environment variables to be set inside launched container, specified as `KEY=VALUE`",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		cli.StringFlag{
+			Name:  "labels, l",
+			Value: "",
+			Usage: "specify labels to be set on launched container",
```
```
Usage: "specify `label`s to be set on launched POD",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		cli.StringFlag{
+			Name:  "limits",
+			Value: "",
+			Usage: "specify resource limits for launched container",
```
```
Usage: "specify `resource limit`s for launched POD",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		cli.StringFlag{
+			Name:  "ports",
+			Value: "",
+			Usage: "specify ports to be forwarded to launched container",
```
```
			Usage: "specify `port`s to be forwarded to launched POD",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.StringSliceFlag{
+			Name:  "device",
+			Usage: "make host devices available inside the container",
```
```
			Usage: "make host `device`s available inside the container",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.StringSliceFlag{
+			Name:  "dns",
+			Usage: "set DNS servers for container",
```
```
		Usage: "set `DNS server`s for POD",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.StringSliceFlag{
+			Name:  "dns-search",
+			Usage: "set DNS search domains for container",
```
```
Usage: "set `DNS search domain`s for POD",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		cli.StringFlag{
+			Name:  "pod",
+			Value: "",
+			Usage: "launch container inside an existing pod",
```
```
			Usage: "launch container inside an `existing pod`",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.BoolFlag{
+			Name:  "privileged",
+			Usage: "launch a privileged container",
```
```
Usage: "launch a privileged POD",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+		},
+		cli.BoolFlag{
+			Name:  "read-only",
+			Usage: "mount root of created container as read only",
```
```
			Usage: "mount root file system of created container as read only",
```

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+func getRandomID() (string, error) {
+	urandom, err := os.Open("/dev/urandom")
+	if err != nil {
+		return "", fmt.Errorf("could not open urandom for reading: %v", err)
```
/dev/urandom

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+	data := make([]byte, 16)
+	count, err := urandom.Read(data)
+	if err != nil {
+		return "", fmt.Errorf("error reading from urandom: %v", err)
```
/dev/urandom

rhatdan commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,553 @@
+	if err != nil {
+		return "", fmt.Errorf("error reading from urandom: %v", err)
+	} else if count != 16 {
+		return "", fmt.Errorf("read too few bytes from urandom")
```
ditto

rhatdan commented on cmd/kpod/main.go:
```diff
@@ -7,12 +7,27 @@ import (
 )
 
+// TODO should share this default with ocid - move inside server, maybe?
+const ociConfigPath = "/etc/ocid/ocid.conf"
```
Yes this needs to be shared with server

rhatdan commented on cmd/kpod/main.go:
```diff
@@ -7,12 +7,27 @@ import (
+		cli.StringFlag{
+			Name:  "config",
+			Value: ociConfigPath,
+			Usage: "path to OCI config file",
```
```
			Usage: "path to `OCI config file`",
```

rhatdan commented:
Missing Man pages and bash completions.

rhatdan commented:
Does kpod launch launch PODs or containers?

rhatdan commented:
docker launch --help 
Should never show "value".

rhatdan commented:
```
./kpod launch --help | grep value
   --image value             image to launch
   --env value               specify environment variables to be set inside launched container, specified as KEY=VALUE
   --labels value, -l value  specify labels to be set on launched container
   --limits value            specify resource limits for launched container
   --ports value             specify ports to be forwarded to launched container
   --mount value             attach mounts on the host to created container
   --device value            make host devices available inside the container
   --dns value               set DNS servers for container
   --dns-search value        set DNS search domains for container
   --pod value               launch container inside an existing pod
```

mheon commented:
@rhatdan Did an initial pass at addressing your comments on the help test. Everything that refers to the primary container effects only a single container, everything referring to a pod affects the entire pod.

Also, hopefully, Travis should be fixed.

nalind commented on cmd/kpod/launch.go:
```diff
@@ -1,16 +1,592 @@
+}
+
+// Generate and hex-encode 128-bit random ID
+func getRandomID() (string, error) {
```
We use github.com/docker/docker/pkg/stringid.GenerateNonCryptoID() for this elsewhere in the server package.

mheon replied:
Good catch, I'll use that implementation instead

cyphar commented on cmd/kpod/launch.go:
```diff
@@ -2,15 +2,568 @@ package main
+		},
+	},
+	Action: func(ctx *cli.Context) error {
+		if ctx.GlobalBool("debug") {
```
Wasn't the original idea that we would refactor out all of the `cri-o` daemon into libraries? I mean, this can be done later but I imagine it's going to be quite ugly to unify the two different bodies of code once we implement everything here.

On the plus side, this code is going to be newer, so maybe we could base the daemon code on this code (after copying it to a library).

mheon commented:
Closing this, working on a new approach
